Broker and Store updates

Dave-Vallance opened the pull request:
Hi Ed!
Here are the suggestions for the issues discussed here:

[Backtrader Community](https://community.backtrader.com/topic/623/anyone-use-backtrader-to-do-live-trading-on-bitcoin-exchange/203)

Dave-Vallance committed:
ccxtbroker: getposition update

Update to fix the getposition() method. It now returns the current open position size and amount.

Dave-Vallance committed:
ccxtstore: Order parsing update

Updates to:
  - Return an order id from create_order. This can then be used to parse the order and return the CCXT standard variables such a 'amount' 
  - Add a fetch_order method that can be used to return the parsed order details.

bartosh commented on backtrader/brokers/ccxtbroker.py:
```diff
@@ -71,18 +74,27 @@ def get_notification(self):
+    def getposition(self, data, clone=True):
+        # return self.o.getposition(data._dataname, clone=clone)
+        pos = self.positions[data._dataname]
+        if clone:
```
What if we already have position when algo started? How we're going to report it?

Dave-Vallance replied:
Hi Ed, 
  Yes - There is still work to do. However, this does get us a step closer I think. 

Regarding the question. It is a good one. The golden sample, aka Daniels own Oanda integration tackles this on init

```
if self.p.use_positions:
            for p in self.o.get_positions():
                print('position for instrument:', p['instrument'])
                is_sell = p['side'] == 'sell'
                size = p['units']
                if is_sell:
                    size = -size
                price = p['avgPrice']
                self.positions[p['instrument']] = Position(size, price)
```

This will get the open positions when we start. Naturally we would need to tweak this to work with the CCXT store.

Dave-Vallance replied:
Hi Ed, 
  Just to expand on above a bit further. The original code for getting positions also would not allow Backtrader to be aware of open positions. This is because the `getposition` method in the store only returns the account value. It needs to return the position size and position entry/avg price for a particular instrument and pass it to the positions dictionary.

Then we can use the `if not self.position` check in Backtrader.   

Snippet from the CCXT store:
```
@retry
    def getposition(self, currency):
        return self.getvalue(currency)
```

The problem I was having yesterday was that I could not find a good unified method for returning open positions in the CCXT library. This is where I started playing with the parse trades method below. I was thinking to get a list of trades and see if any of them were open. I did not resolve this item though.

bartosh commented on backtrader/stores/ccxtstore.py:
```diff
@@ -116,10 +117,19 @@ def cancel_order(self, order_id):
 
+
+    @retry
+    def parse_trade(self, trade, market=None):
```
Where this method is used?

Dave-Vallance replied:
Sorry, this is not currently used. I think I was testing some ideas and forgot to remove it. 

I can recommit with this removed if you like?

bartosh replied:
yes, please update this commit and push again.

Dave-Vallance committed:
Cancel_order Fix and Pass Params

Fix for TypeError raised when trying to cancel an order. 

Fix for Passing of parameters dict to create order

Dave-Vallance committed:
_Submit: Passing Params Fix

Added extra line to check for existance of exchange specific parameters.

Dave-Vallance committed:
ccxtbroker: Order Filling

Update to track filling of orders. This update appends all submitted orders to a list. These are considered open orders until they are confirmed to be filled. The brokers next() method to get a list of all closed orders and it checks if any orders from the open orders list are in there. If so, we know the order has been completed.

Dave-Vallance committed:
ccststore: Remove parse_trade add fetch_closed_orders

Update to remove unused parse trade method

Added fetch_closed_orders to work will order filling

Dave-Vallance commented:
Hi Ed, 
  Just pushed some new updates. I removed the `parse_trade()` method. 

I also have a proposal in the code for how to track when orders are filled. It doesn't feel too elegant but it seems to work in my test environment. 

Let me know if you what you think about it and if you have any questions. Maybe you have a better idea? 

BTW - To parse stop loss orders (if you decide to test them) and you use bitmex, you will need to update to the latest development version of CCXT. I reported an issue with parsing orders when orders are created using extra non-unified parameters. The stop loss is one such case where a `stopPx` level is set rather than `price`

EDIT: I think there is an issue with removing the order from my open_orders list. I will look into it tomorrow.

Dave-Vallance committed:
ccxtstore: Remove Fetch Closed Trades

No longer needed

Dave-Vallance committed:
ccxtbroker: order filling improvments

improvements to the detection of when orders are filled. 

fix for canceling orders which have already been filled.

Dave-Vallance commented:
Hi Ed, 
Updates from today. 

I have done some more work on detecting order status and fills. 

Using fetch_closed_orders worked well for one day then resulted in an issue on bitmex. Essentially, the returned list for closed orders was limited and did not return the most recent history. Further, the `since` parameter did not affect the returned results. Not sure if that is a CCXT issue. I was able to work around this by passing exchange specific parameters to the call but that is not a satisfactory solution. 

To work around this, I am calling `fetch_order` on each call of next to check the status one by one. It is more reliable but the downside is that it can result in more API calls if you have lots of open orders. 

I have now been able to successfully detect fills reliably for market and stop loss orders.

Dave-Vallance committed:
fetch_ohlc_params: Update

Update to allow passing of a params dict when creating the data feed.

bartosh commented:
Hi Dave,

Now I can see 9 commits in this PR. They're quite hard to review. Can you squash them into smaller amount of commits? It would be great if you can split your changes into several PRs if they're independent from each other. That would greatly simplify review process for me. Thanks.

Dave-Vallance commented:
Hi Ed,
  Sure, I have admittedly gone a bit crazy in the past couple of days trying to get a project off the ground. Let me review and try to split things up as you have suggested.
